### PR TITLE
Fixes #32841 - Host registration - :rhsm_base_url helper

### PR DESCRIPTION
**Pull request by stejskalleos**
For https://github.com/theforeman/foreman/pull/8563

### CODE REVIEW

File: app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb [hunk 1]
```diff
@@ -24,6 +24,25 @@ def host_setup_extension
...
+
+      def rhsm_base_url
+        smart_proxy = if params[:url]
+          SmartProxy.find_by(url: params[:url])
+        else
+          SmartProxy.pulp_primary
+        end
+
+        fail Foreman::Exception.new('Smart proxy content source not found!') unless smart_proxy
+        URI.join(smart_proxy.url, 'pulp/content')
```
Comment (ekohl) [inline app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb:44]: I'd expect RHSM on `/rhsm`

**Review by ekohl — CHANGES_REQUESTED** [review]
Oh, another complication: the RHSM port is 8443 on Pulp mirrors. I really think https://github.com/Katello/katello/pull/9413 (which implements it as a function on the Smart Proxy) is better. For reference, the relevant bit:
```
      def rhsm_url
        # TODO: We shouldn't relate this to the Pulp feature and also expose it
        # as a setting
        if pulp_mirror?
          "https://#{URI.parse(url).host}:8443/rhsm"
        elsif pulp_primary?
          "https://#{URI.parse(url).host}/rhsm"
        end
      end
```

File: app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb (hunk 1 above)
Comment (stejskalleos) [reply app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb:44]: When I install subscription manager with `katello_ca_consumer.rpm` the `/etc/rhsm/rhsm.conf` config file is set like this:
```
[rhsm]
# Content base URL:
baseurl = https://centos7-katello-devel-stable.example.com/pulp/content/
```

So I thought that it should be same

Comment (ekohl) [reply app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb:44]: That's the (Pulp) content URL, not RHSM itself. I'd consider the RHSM URL the basis for `subscription-manager config --server.hostname "$HOSTNAME" --server.port "$PORT" --server.prefix "$PREFIX"`.

Also, RHSM is always HTTPS while the Smart Proxy URL may be HTTP.

Comment (stejskalleos) [reply app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb:44]: Hmm, I'm getting super confused from all of the URLs

Right now there are:
* Foreman URL
* Smart Proxy URL
* Primary Pulp content URL
* Mirror pulp content URL
* RHSM URL

If I want to configure subscription-manager, what URLs should I use for these options:

```
subscription-manager config \
      --server.hostname="" \
      --server.port="" \
      --server.prefix="/rhsm" \
      --rhsm.baseurl=""
```

Comment (ekohl) [reply app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb:44]: It is my understanding (based on https://github.com/theforeman/puppet-foreman_proxy_content/blob/master/lib/puppet/provider/rhsm_reconfigure_script/rhsm_reconfigure_script.rb) that `--server.*` is the RHSM URL. `--rhsm.baseurl` is the Pulp content URL.

Comment (ehelms) [reply app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb:44]: Correct.

`server.hostname`: is the hostname subscription-manager should communicate with when making API calls
`server.port`: is the port subscription-manager should communicate with when making API calls
`server.prefix`: is a URL prefix that subscription-manager should talk to for all its API communication back to the server, in our case that is always `/rhsm`
`rhsm.baseurl`: is the location at which all content is located, this is prepended to all content paths when constructing the `redhat.repo` file

For example, if Candlepin knows a repo path is `/Default_Organization/Library/Mv_CV/repo_a/` and the `rhsm.baseurl` is `https://mycapsule.com/pulp/content` then the redhat.repo will get configured with:

```
baseurl=https://mycapsule.com/pulp/content/Default_Organization/Library/My_CV/repo_a
```

File: app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb [hunk 2]
```diff
@@ -24,6 +24,25 @@ def host_setup_extension
...
+        urls = super
+        urls[:rhsm_base_url] = rhsm_base_url
+        urls
+      end
+
+      private
+
+      def rhsm_base_url
+        smart_proxy = if params[:url]
+          SmartProxy.find_by(url: params[:url])
```
Comment (ehelms) [inline app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb:38]: If we are being technically correct, we should ensure that the found smart proxy has the Pulp 3 feature and throw an error if not.

File: app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb [hunk 3]
```diff
@@ -24,6 +24,32 @@ def host_setup_extension
...
+        fail Foreman::Exception, N_('Smart proxy content source not found!') unless @smart_proxy
+        fail Foreman::Exception, N_('Pulp 3 is not enabled on Smart proxy!') unless @smart_proxy.pulp3_enabled?
+      end
+
+      def rhsm_url
+        URI(@smart_proxy.rhsm_url)
+      end
+
+      def pulp_content_url
+        URI.join(@smart_proxy.url, '/pulp/content')
```
Comment (ekohl) [inline app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb:51]:
```suggestion
        @smart_proxy.setting(SmartProxy::PULP3_FEATURE, 'content_app_url')
```

**Review by ekohl — CHANGES_REQUESTED** [review]

**Review by jturel — COMMENTED** [review]
Left two comments on code style but testing in conjunction with the Foreman PR works well. Here's the resulting bits from my rhsm.conf:

```
hostname = centos7-katello-devel.jturel.example.com
prefix = /rhsm
port = 443
[rhsm]
# Content base URL:
baseurl = https://centos7-katello-devel.jturel.example.com/pulp/content

```

File: app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb [hunk 4]
```diff
@@ -24,6 +24,29 @@ def host_setup_extension
...
+        super.merge(rhsm_url: rhsm_url, pulp_content_url: pulp_content_url)
+      end
+
+      private
+
+      def find_smart_proxy
+        @smart_proxy ||= params[:url] ? SmartProxy.find_by(url: params[:url]) : SmartProxy.pulp_primary
+
+        fail Foreman::Exception, N_('Smart proxy content source not found!') unless @smart_proxy
+        fail Foreman::Exception, N_('Pulp 3 is not enabled on Smart proxy!') unless @smart_proxy.pulp3_enabled?
```
Comment (ekohl) [inline app/controllers/katello/concerns/api/v2/registration_controller_extensions.rb:40]: A common pattern for this is
```suggestion
      def smart_proxy
        @smart_proxy ||= begin
          proxy = params[:url] ? SmartProxy.find_by(url: params[:url]) : SmartProxy.pulp_primary

          fail Foreman::Exception, N_('Smart proxy content source not found!') unless proxy
          fail Foreman::Exception, N_('Pulp 3 is not enabled on Smart proxy!') unless proxy.pulp3_enabled?
          
          proxy
        end
```

Then you can just call `smart_proxy.rhsm_url` without first having to call `find_smart_proxy`. Not having to rely on side effects is IMHO a good pattern.

**Review by jturel — APPROVED** [review]
APJ

**Review by ekohl — APPROVED** [review]

**Comment by jturel** [comment]
[test katello]

**Pull request merged**